File: src/drivers/src/flare_listener.cpp
#include <vector>
#include <string>

#include <opencv2/opencv.hpp>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/bool.hpp"

#include "jetson_gpio.hpp"

class FlareListener : public rclcpp::Node
{

public:
    FlareListener() : Node("flare_listener")
    {
        RCLCPP_INFO(this->get_logger(), "Starting LED listener...");

        // parse parameters
        std::string flare_topic;
        int flare_pin;
        cv::FileStorage fs;
        fs.open("/home/user/ws/src/config/config.yaml", cv::FileStorage::READ);
        fs["actuators"]["flare"]["topic"] >> flare_topic;
        fs["actuators"]["flare"]["pin"] >> flare_pin;
        fs.release();

        // Initialize led tape
        this->flare_gpio = JetsonGPIO(flare_pin, JetsonGPIO::Direction::OUTPUT, JetsonGPIO::Value::LOW);        

        // initialize subscribers
        this->flare_subscriber = this->create_subscription<std_msgs::msg::Bool>( flare_topic, 10, std::bind(&FlareListener::flare_callback, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "Flare listener has been started.");
    }

private:
    JetsonGPIO flare_gpio;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr flare_subscriber;

    void flare_callback(const std_msgs::msg::Bool::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "Received message: %d", msg->data);
        if (msg->data)
        {
            this->flare_gpio.setValue(JetsonGPIO::Value::HIGH);
        } else 
        {
            this->flare_gpio.setValue(JetsonGPIO::Value::LOW);
        }
    }
};

int main(int argc, char const *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<FlareListener>());
    rclcpp::shutdown();

    return 0;
}

File: src/drivers/scripts/flare_listener.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import cv2 

from std_msgs.msg import Bool

import Jetson.GPIO as GPIO

class FlareListener(Node):

    def __init__(self):
        super().__init__('flare_listener')
        self.get_logger().info('Starting LED listener...')

        # Parse parameters
        fs = cv2.FileStorage("/home/user/ws/src/config/config.yaml", cv2.FileStorage_READ)
        flare_config = fs.getNode("actuators").getNode("flare")
        self.topic = flare_config.getNode("topic").string()
        self.pin = int(flare_config.getNode("pin").real())
        fs.release()

        # Start jetson GPIO
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(self.pin, GPIO.OUT, initial=GPIO.LOW)

        self.flare_subscriber = self.create_subscription(Bool, self.topic, self.flare_callback, 10)

        self.get_logger().info("Initialize flare subscriber")

    def flare_callback(self, msg):

        # self.get_logger().info(f"Received msg: {msg.data}")

        value = GPIO.HIGH if msg.data else GPIO.LOW
        GPIO.output(self.pin, value)

if __name__ == "__main__":
    rclpy.init(args=None)

    flare_listener = FlareListener()

    rclpy.spin(flare_listener)

    flare_listener.destroy_node()
    rclpy.shutdown()
